### ISSUE TITLE
Distribute training problem

### ISSUE THREAD

**Issue by fanlu** [vrenkens/nabu#31]
Hi Vincent,
I use mode=single_machine and mode=multi_machine with cluster file,
but the GPU card is not efficient,and There are only one gpu card work(GPU-Util>0)
but the both two gpu card's memory is full.
I remove the ssh tunnel code, Are there any other configurations?
Thu Jun 14 16:33:22 2018
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.81                 Driver Version: 384.81                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P40           Off  | 00000000:02:00.0 Off |                    0 |
| N/A   39C    P0    56W / 250W |  21813MiB / 22912MiB |     20%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla P40           Off  | 00000000:03:00.0 Off |                    0 |
| N/A   38C    P0    51W / 250W |  21779MiB / 22912MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0   2829388      C   python                                     21803MiB |
|    1   2830147      C   python                                     21769MiB |
+-----------------------------------------------------------------------------+

**Comment by vrenkens** [comment]
This really depends on a lot of factors. Nvidia-smi also only gives you momentary information, for the full view you would need statistics over a longer timespan. 

What is the performance on a single machine in terms of time per step and how does it compare to the distributed case?

**Comment by fanlu** [comment]
I use `watch -n 1 nvidia-smi` to statistic GPU-Util every second, and one of the GPU card GPU-Util is also 0%.
And the worker log print alternately.
e.g.
WORKER 0: step 843/339700 loss: 143.967102, learning rate: 0.000994 
         time elapsed: 5.798077 sec
         peak memory usage: 21526/22600 MB
WORKER 1: step 844/339700 loss: 141.791336, learning rate: 0.000994 
         time elapsed: 6.255881 sec
         peak memory usage: 21526/22600 MB
WORKER 0: step 845/339700 loss: 137.241943, learning rate: 0.000994 
         time elapsed: 6.195476 sec
         peak memory usage: 21526/22600 MB
WORKER 1: step 846/339700 loss: 134.677200, learning rate: 0.000994 
         time elapsed: 4.318811 sec
         peak memory usage: 21526/22600 MB
WORKER 1: step 848/339700 loss: 100.846954, learning rate: 0.000994 
         time elapsed: 3.668412 sec
         peak memory usage: 21526/22600 MB
WORKER 0: step 847/339700 loss: 124.031097, learning rate: 0.000994 
         time elapsed: 4.787115 sec
         peak memory usage: 21526/22600 MB
And The GPU-Util can grow up to 90% only when evaluation.
And I use one machine to simulate distributed case.

**Comment by fanlu** [comment]
@vrenkens 
the function def _device in `trainer.py`
```
 device = tf.train.replica_device_setter(
                ps_tasks=num_servers,
                ps_strategy=ps_strategy)
```
you don't add  param `worker_device='/job:worker/task:%d' % self.args.task_index` and param `cluster=cluster`, why?

**Comment by vrenkens** [comment]
It has been a while since I coded this, so I'm a bit rusty. Sorry for that. I think you are right these specifications should be there. Could you try with

```
device = tf.train.replica_device_setter(
    ps_tasks=num_servers,
    ps_strategy=ps_strategy,
    worker_device='/job:worker/task:%d' % self.task_index,
    cluster=cluster )
``` 

and let me know what happens?

**Comment by fanlu** [comment]
All gpu GPU-Util>0 when I set param worker_device, cluster and run training.
So it's solved.

**Comment by vrenkens** [comment]
Thanks!

**Comment by AzizCode92** [comment]
I still have the same issue, I have 3 gpus in one single machine and I am trying to use all of them.
Unfortunately Only one gpu is used and the two others have been allocated memory but zero percentage of activity.

+-----------------------------------------------------------------------------+
| NVIDIA-SMI 387.26                 Driver Version: 387.26                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 00000000:00:19.0 Off |                    0 |
| N/A   62C    P0    64W / 149W |  10896MiB / 11439MiB |     21%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K80           Off  | 00000000:00:1A.0 Off |                    0 |
| N/A   49C    P0    69W / 149W |  10873MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla K80           Off  | 00000000:00:1B.0 Off |                    0 |
| N/A   55C    P0    57W / 149W |  10873MiB / 11439MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+


Any hints?
thanks
